packages/modules/notification/src/services/notification-provider.ts

```diff
@@ -0,0 +1,64 @@
+import { Constructor, DAL, NotificationTypes } from "@medusajs/types"
+import { ModulesSdkUtils } from "@medusajs/utils"
+import { MedusaError } from "medusa-core-utils"
+import { NotificationProvider } from "@models"
+import { NotificationProviderRegistrationPrefix } from "@types"
+
+type InjectedDependencies = {
+  notificationProviderRepository: DAL.RepositoryService
+  [
+    key: `${typeof NotificationProviderRegistrationPrefix}${string}`
+  ]: NotificationTypes.INotificationProvider
+}
+
+export default class NotificationProviderService extends ModulesSdkUtils.internalModuleServiceFactory<InjectedDependencies>(
+  NotificationProvider
+) {
+  protected readonly notificationProviderRepository_: DAL.RepositoryService<NotificationProvider>
+  // We can store the providers in a memory since they can only be registered on startup and not changed during runtime
+  protected providersCache: Map<string, NotificationProvider>
+
+  constructor(container: InjectedDependencies) {
+    super(container)
+    this.notificationProviderRepository_ =
+      container.notificationProviderRepository
+  }
+
+  protected retrieveProviderRegistration(
+    providerId: string
+  ): NotificationTypes.INotificationProvider {
+    try {
+      return this.__container__[
+        `${NotificationProviderRegistrationPrefix}${providerId}`
+      ]
+    } catch (err) {
+      throw new MedusaError(
+        MedusaError.Types.NOT_FOUND,
+        `Could not find a notification provider with id: ${providerId}`
+      )
+    }
+  }
+
+  async getProviderForChannel(
+    channel: string
+  ): Promise<NotificationProvider | undefined> {
+    if (!this.providersCache) {
+      const providers = await this.notificationProviderRepository_.find()
+      this.providersCache = new Map(
+        providers.flatMap((provider) =>
+          provider.channels.map((c) => [c, provider])
+        )
+      )
+    }
+
+    return this.providersCache.get(channel)
```

Comment:
q: can't we have multiple providers for a single channel?

Reply:
@adrien2p I think it is better if we don't for the time being. We can always loosen this restriction, but for now I'd rather not push provider selection to the caller, as it introduces unnecessary complexity for almost all of the users.

We already discussed this in the RFC if you want to check that out

Reply:
I do agree with that, but we might as well throw an error if it occurs by mistakes? wdyt?

Reply:
@adrien2p yup agreed, I'll add a check in the loader :+1